File: 2019/Thang02_2018/24_02_2019/swindow.cpp
#include <iostream>
#include <stdio.h>
#include <string.h>

using namespace std;
///////

const
    int maxn=1e6+100;

int na, nb, nc, k;
char a[maxn], b[maxn];
int z[maxn*2], d[maxn*2];
char c[maxn*2];
int res;
///////

void enter()
{
    scanf("%[^\n]\n", a); na=strlen(a);
    scanf("%[^\n]\n", b); nb=strlen(b);
    scanf("%d", &k);
}
///////

void gan(char x[], char y[])
{
    int ny=strlen(y);
    for (int i=0; i<ny; i++) x[i]=y[i];
    x[ny]='\0';
}

void noi(char x[], char y[])
{
    int nx=strlen(x), ny=strlen(y);
    for (int i=0; i<ny; i++) x[nx+i]=y[i];
    x[nx+ny]='\0';
}

void process()
{
    res=0;
    if (nb>k) return;

    gan(c, b);
    noi(c, "@"); noi(c, a);
    nc=strlen(c);
    z[0]=0;

    int l=0, r=0;
    for (int i=1; i<nc; i++)
    {
        z[i]=0;
        if (i<=r) z[i]=min(r-i+1, z[i-l]);
        while (i+z[i]<nc && c[i+z[i]]==c[z[i]]) z[i]++;
        if (i+z[i]-1>r)
        {
            l=i; r=i+z[i]-1;
        }
        if (z[i]>=nb)
        {
            d[max(0, i-k+nb)]++;
            d[i+1]--;
        }
    }

    for (int i=0; i<nc-k+1; i++)
    {
        if (i>0) d[i]+=d[i-1];
        if (d[i]>0) res++;
    }
    cout << res;
}

int main()
{
    freopen("swindow.inp", "r", stdin);
    freopen("swindow.out", "w", stdout);
    enter();
    process();
}

File: 2018/Thang10-2018/12-10-2018/swindow.pas
Uses    Math;

Const   ginp='swindow.inp';
        gout='swindow.out';

Var     n,m,k,res:longint;
        a,b:ansistring;
        z:array[0..2000010] of longint;

Procedure Enter;
Begin
        readln(a);
        readln(b); m:=length(b);
        a:=b+'$&'+a; n:=length(a);
        read(k);
End;

Procedure Process;
Var     i,l,r,pos:longint;
Begin
        res:=0; pos:=0;
        if k<m then exit;
        z[1]:=0; l:=0; r:=0;
        for i:=2 to n do
        begin
                if i<=r then z[i]:=min(r-i+1,z[i-l+1]);
                while (i+z[i]<=n) and (a[1+z[i]]=a[i+z[i]]) do inc(z[i]);
                if r<i+z[i]-1 then
                begin
                        l:=i; r:=i+z[i]-1;
                end;
                if z[i]=m then
                begin
                        pos:=max(pos+1,max(m+3,i-k+m));
                        for pos:=pos to min(i,n-k+1) do inc(res);
                end;
        end;
End;

Procedure Answer;
Begin
        write(res);
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Enter;
        Process;
        Answer;
        Close(input); Close(output);
End.
